src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/TypeOperations.cs

```diff
@@ -152,8 +153,12 @@ public GetValueOrDefault(MethodInfo mi)
                 _defaultValueType = mi.ReturnType;
             }
 
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2077:UnrecognizedReflectionPattern",
+                Justification = "_defaultValueType is a ValueType, so it will always have a default constructor.")]
```

Comment:
This is not strictly correct. Structs can actually declare a default constructor in IL and can have some code in it. If illink were to strip the default constructor, the CreateInstance call below would still work, but it would do a different thing.

I'm not sure what the intent of this code here is (whether it wants an all-zero default instance and calling CreateInstance is actually a bug (should be using GetUninitializedObject instead), or it wants to run the default constructor if there is any, in which case the suppression is incorrect.

C# actually wanted to add support for these default constructors in the past, but it was pulled last minute because of a runtime bug. It might come back. https://github.com/dotnet/runtime/issues/6536

Reply:
> C# actually wanted to add support for these default constructors in the past, but it was pulled last minute because of a runtime bug. It might come back.

I think it's likely to in C# 10.

Reply:
> I'm not sure what the intent of this code here is 

My understanding is the intent is to simulate calling `Nullable<>.GetValueOrDefault()`. See:

https://github.com/dotnet/runtime/blob/e2099f62f1ee28241f857edaf0ef7420018d11d8/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs#L2183-L2188

Looking at the `Nullable<T>` code, the only way `Nullable<T>` doesn't have a value (and `frame.Peek() == null` is true) is if the nullable was created by getting the `default(Nullable<T>)`.

https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/System.Private.CoreLib/src/System/Nullable.cs#L16-L47

In that case, the `internal T value` will have been created by `default(T)`. So no constructor will have run.

Given this, I think the more appropriate change here is to use `GetUninitializedObject`. We will still need the same suppression however, since `GetUninitializedObject` is annotated the same way `Activator.CreateInstance` is.

Maybe the best thing to do here is to open an issue for this and not change System.Linq.Expressions during annotation.

Reply:
Opened https://github.com/dotnet/runtime/issues/47647 for this.